### DISCUSSION TITLE
TextLayer rendering on same pane / z-fighting (8.5.0-alpha.x)

### DISCUSSION THREAD

**Discussion by rammoozz** [visgl/deck.gl#5672] [Q&A]
#### Description
TextLayer rendering on same pane, z-fighting (8.5.0-alpha.x)

#### Expected Behavior
TextLayer stacks per 8.4

#### Repro Steps
https://codesandbox.io/s/deckgl-forked-ok3n8?file=/src/DeckGLOverlay.js
```
 new TextLayer({
          data: [
            {
              getPixelOffset: [0, 3],
              position: [-102.552784, 23.634501],
              text: "AAAAAAAAAAAAAAAAAAAAAAAAA"
            },
            {
              getPixelOffset: [0, -3],
              position: [-102.552784, 23.634501],
              text: "BBBBBBBBBBBBBBBBBBBBBBBBBB"
            }
          ],

          backgroundPadding: [8, 4],
          getColor: [180, 0, 0],
          backgroundColor: [0, 0, 0],
          getPixelOffset: (d) => d.getPixelOffset,
          getText: (d) => d.text
        })
```
#### Environment
 - Framework Version: 
- @deck.gl/core: 8.5.0-alpha.1
- @deck.gl/google-maps: 8.5.0-alpha.1
- @deck.gl/layers: 8.5.0-alpha.1

![image](https://user-images.githubusercontent.com/5125481/111852696-597eac80-88ee-11eb-9f7c-9bbc35685da4.png)

**Comment by Pessimistress** [comment]
This is one of the breaking changes from adding the new TextLayer styling features. Backgrounds are now rendered with a separate layer from the characters. You can work around this by adding a small z to each position.

**Reply by rammoozz** [reply]
> This is one of the breaking changes from adding the new TextLayer styling features. Backgrounds are now rendered with a separate layer from the characters. You can work around this by adding a small z to each position.

Thank you, when adding a z to each position, it seems to still overlap at high zooms, _is this intended?_


![image](https://user-images.githubusercontent.com/59570712/113514560-61209100-953d-11eb-935d-8955b87fb5e9.png)

changing it to a difference of ~3 fixes the original issue found until max zoom, _should this be in the documentation?_

![image](https://user-images.githubusercontent.com/59570712/113514611-b2308500-953d-11eb-8204-2d58d8ec1ad4.png)
![image](https://user-images.githubusercontent.com/5125481/113524901-c301ea80-957f-11eb-89e1-3ca0c520bedf.png)

The problem with using larger numbers than that is the 'size' is impacted as a side effect. How do you recommend dealing with this at scale? (A large number of TextLayer in one area, even with a random z 1-10 there's bound to be some that stack)

**Comment by Pessimistress** [comment]
> changing it to a difference of ~3 fixes the original issue found until max zoom, should this be in the documentation?

The z offset that is needed depends on multiple viewport/WebGL statuses. You may want to try adjust/disable [getPolygonOffset](https://deck.gl/docs/api-reference/core/layer#getpolygonoffset) for more predictable results.

**Reply by rammoozz** [reply]
I am having issues scaling the solution by adding random numbers.
- using the index, goes up to N, issues with zooming.
- random number between 0-20, with N text layers, there's going to be conflicts.

My attempts to set getPolygonOffset also had no impact, even via sublayerprops, when adding it to the 'background' sublayerprop, the characters disappear, but the 'characters' sublayer ignores it

https://codesandbox.io/s/deckgl-forked-zxsfw?file=/src/DeckGLOverlay.js

![image](https://user-images.githubusercontent.com/5125481/114706110-932cb280-9cf6-11eb-9c23-28c226492712.png)

**Reply by Pessimistress** [reply]
`getPolygonOffset` is not called per object. It is called once before the layer is drawn to set [glPolygonOffset](https://www.khronos.org/registry/OpenGL-Refpages/es2.0/xhtml/glPolygonOffset.xml).

To disable it you need:

`getPolygonOffset: null`

**Reply by rammoozz** [reply]
> To disable it you need:
> 
> `getPolygonOffset: null`

I have that property but it doesn't seem to change anything, did I include it in the right spot?


![image](https://user-images.githubusercontent.com/5125481/115115151-b5c20400-9f60-11eb-8c56-b9b8a5462c22.png)

**Reply by Pessimistress** [reply]
I don't see the random z in your positions. Can you update your sandbox to include all the solutions that we've talked about?

**Reply by rammoozz** [reply]
[Updated](https://codesandbox.io/s/deckgl-forked-zxsfw?file=/src/DeckGLOverlay.js), added back the random z to position prop: 
The random z (0-20) helps but it does not scale well with a lot of data.

With both `getPolygonOffset:null` and random z

![image](https://user-images.githubusercontent.com/5125481/115129959-12e9a400-9fb9-11eb-8589-783539742998.png)

**Reply by Pessimistress** [reply]
Just remove the `Math.round` for z

**Reply by rammoozz** [reply]
Better...there's still conflicts, increasing > 20 helps, then there's issues when zooming in 

![image](https://user-images.githubusercontent.com/5125481/115130012-b5098c00-9fb9-11eb-8420-28b2dc17e970.png)

**Reply by rammoozz** [reply]
max zoom also is unreadable:

![image](https://user-images.githubusercontent.com/5125481/115130027-ec783880-9fb9-11eb-9a0e-5e57716f46d1.png)

**Reply by Pessimistress** [reply]
You need bigger z offsets as the zoom gets smaller. You can do this by changing the `modelMatrix` prop, something along the lines of:

```js
import {Matrix4} from 'math.gl';

new TextLayer({
  getPosition: d => [d.longitude, d.latitude, Math.random()],
  modelMatrix: new Matrix4().scale([1, 1, 2**(12-zoom)])
})
```

And rerender when zoom changes.

Realistically though, why would you draw so many overlapping texts to begin with? Users won't be able to read them either way.

**Reply by rammoozz** [reply]
Thanks! I'll try.

> Realistically though, why would you draw so many overlapping texts to begin with? Users won't be able to read them either way.

The one top would always be readable though? As long as they don't collide.

**Reply by rammoozz** [reply]
[Updated](https://codesandbox.io/s/deckgl-forked-zxsfw?file=/src/DeckGLOverlay.js):
Removing the 1-20 helps with zooming in and size, but the collision is still visible

![image](https://user-images.githubusercontent.com/5125481/115130352-36aee900-9fbd-11eb-96da-e02237765664.png)

![image](https://user-images.githubusercontent.com/5125481/115130374-4f1f0380-9fbd-11eb-8596-0fc5bd6d521b.png)

**Reply by Pessimistress** [reply]
You are not updating the layer when zoom changes. You need to call `overlay.setProps({layers})` on each render.

**Reply by rammoozz** [reply]
I've [updated ](https://codesandbox.io/s/deckgl-forked-zxsfw?file=/src/DeckGLOverlay.js)the code 

![image](https://user-images.githubusercontent.com/5125481/115130774-4419a280-9fc0-11eb-85d3-a8a2e05bfadf.png)

**Reply by rammoozz** [reply]
I believe this is related, happens on max zoom:

![Untitled Project](https://user-images.githubusercontent.com/5125481/115315950-5de4f200-a146-11eb-8472-efb702de622b.gif)

**Reply by Pessimistress** [reply]
There is still a fair amount of collision. `2 ** (12- zoom)` is just showcasing the rough relationship between z scale and zoom, you need to play with the actual multiplier to set the labels apart. Also, replacing `Math.random()` with a deterministic number is probably a good idea.

**Reply by rammoozz** [reply]
Thank you for the help!

I've changed the z to deterministic numbers, it looks **much** better with 10,000 TextLayers, [code](https://codesandbox.io/s/deckgl-forked-dbtj5?file=/src/DeckGLOverlay.js).

```
const zindex = rng.next(z) * 0.09
```

![localhost_3000_ (8)](https://user-images.githubusercontent.com/5125481/115463560-f2f4f300-a1f9-11eb-933e-e9667d1c5754.png)

There are instances of collision:

![image](https://user-images.githubusercontent.com/5125481/115465059-01dca500-a1fc-11eb-9f6a-15c7033568df.png)

At high zooms with 100,000 TextLayers, they start to stack and obscure, I'm assuming this something that can be addressed by adjusting the numbers as well.


![localhost_3000_ (9)](https://user-images.githubusercontent.com/5125481/115463561-f38d8980-a1f9-11eb-962a-3ecfdfc778cc.png)

![localhost_3000_ (10)](https://user-images.githubusercontent.com/5125481/115463562-f38d8980-a1f9-11eb-829b-4f2175c4ee4d.png)


Should this be added to the documentation?

**Reply by Pessimistress** [reply]
If you don't have any 3D data to render you can set the view to `new MapView({orthographic: true})` to hide the z shift.

**Reply by rammoozz** [reply]
I am using  [GoogleMapOverlays ](https://deck.gl/docs/api-reference/google-maps/google-maps-overlay#constructor-1), It looks like `views` is not a supported prop.
